pkg/controller/certificates/signer/signer.go

```diff
@@ -89,64 +91,72 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
 }
 
 func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
-	signerName := ""
+	// Ignore unapproved requests
+	if !certificates.IsCertificateRequestApproved(csr) {
+		return nil
+	}
+
 	// Special-case handling here in case we are talking to an older apiserver.
+	signerName := ""
 	if csr.Spec.SignerName != nil {
 		signerName = *csr.Spec.SignerName
 	} else {
 		signerName = capihelper.DefaultSignerNameFromSpec(&csr.Spec)
 	}
 
-	// Only handle CSRs with the specific known signerNames.
-	if !isOneOf(signerName,
-		capi.KubeletServingSignerName,
-		capi.KubeAPIServerClientKubeletSignerName,
-		capi.KubeAPIServerClientSignerName,
-		capi.LegacyUnknownSignerName,
-	) {
+	// Fast-path to avoid any additional processing if the CSRs signerName does
+	// not have a 'kubernetes.io/' prefix.
+	if !strings.HasPrefix(signerName, "kubernetes.io/") {
 		return nil
 	}
-	if !certificates.IsCertificateRequestApproved(csr) {
+
+	x509cr, err := capihelper.ParseCSR(csr.Spec.Request)
+	if err != nil {
+		return fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+	}
+	if !requestValidForSignerName(x509cr, csr.Spec.Usages, signerName) {
 		return nil
 	}
-	csr, err := s.sign(csr)
+	cert, err := s.sign(x509cr, csr.Spec.Usages)
 	if err != nil {
 		return fmt.Errorf("error auto signing csr: %v", err)
 	}
+	csr.Status.Certificate = cert
 	_, err = s.client.CertificatesV1beta1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})
 	if err != nil {
 		return fmt.Errorf("error updating signature for csr: %v", err)
 	}
 	return nil
 }
 
-func (s *signer) sign(csr *capi.CertificateSigningRequest) (*capi.CertificateSigningRequest, error) {
-	x509cr, err := capihelper.ParseCSR(csr.Spec.Request)
-	if err != nil {
-		return nil, fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
-	}
-
+func (s *signer) sign(x509cr *x509.CertificateRequest, usages []capi.KeyUsage) ([]byte, error) {
 	currCA, err := s.caProvider.currentCA()
 	if err != nil {
 		return nil, err
 	}
 	der, err := currCA.Sign(x509cr.Raw, authority.PermissiveSigningPolicy{
 		TTL:    s.certTTL,
-		Usages: csr.Spec.Usages,
+		Usages: usages,
 	})
 	if err != nil {
 		return nil, err
 	}
-	csr.Status.Certificate = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: der})
-	return csr, nil
+	return pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: der}), nil
 }
 
-// isOneOf will return true if 's' is one of 'opts'
-func isOneOf(s string, opts ...string) bool {
-	for _, o := range opts {
-		if s == o {
-			return true
-		}
+func requestValidForSignerName(req *x509.CertificateRequest, usages []capi.KeyUsage, signerName string) bool {
+	// Only handle CSRs with the specific known signerNames.
+	switch signerName {
+	case capi.KubeletServingSignerName:
+		return capihelper.IsKubeletServingCSR(req, usages)
+	case capi.KubeAPIServerClientKubeletSignerName:
+		return capihelper.IsKubeletServingCSR(req, usages)
```

Comment:
@liggitt if the CSR does not exactly match, what should the signing controller do? It could either:

1) override the values
2) reject the request/mark it as invalid somehow (we don't have this option yet)
3) display a warning to the user
4) do nothing at all

Right now with this change we do (4), and simply `return nil`.

Reply:
Adding (3) by simply firing an Event and returning nil could work, but has the potential to cause a lot of events as we have no way to recognise the terminal state of the CR.

(2) is the best solution long-term for external signers IMO, but probably needs consideration in the KEP due to the awkward design of the CSR `conditions` structure.

(1) is an option too - it'd require duplicating/shuffling around some of this checking logic to allow for usages to be different to what's requested by the user. It may be non-obvious though, and it *isn't* something that's been done in the past (old logic *requires* usages to be specified as-per policy). Given it's a net-addition and goes beyond what was described in the KEP changes however, I feel like it might be a bit brash to add it in here.

Reply:
Do 4 for now. I'm looking into how to let signers indicate permanent failure in the CSR status